### PR TITLE
[DRAFT][dredd] Introduce dredd test for TensorShape

### PR DESCRIPTION
**Pull request by icodo98**
Introduce new dredd-rule for tensors shape.

ONE-DCO-1.0-Signed-off-by: JuYoung Lee rsb98759@gmail.com

### CODE REVIEW

**Comment by icodo98** [comment]
related : #14050

File: res/TensorFlowLiteRecipes/Pad_002/test.rule [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# To check if dynamic dimension properly infered
+
+RULE    "VERIFY_FILE_FORMAT"      $(verify_file_format) '=' 1
+
+RULE    "PAD_EXIST"               $(op_count PAD) '=' 1
+RULE    "PAD_SHAPE"               $(tensor_shape ofm) '=' -1572
```
Comment (qsunki) [inline res/TensorFlowLiteRecipes/Pad_002/test.rule:6]: `-1 5 7 2` is more readable than `-1572`. My suggestion is to format it as `$(tensor_shape ofm) = '-1 5 7 2'`, with the appropriate adjustments in the `DumpTensorShape::run` function in `dump.cpp`.

Comment (icodo98) [reply res/TensorFlowLiteRecipes/Pad_002/test.rule:6]: > `-1 5 7 2` is more readable than `-1572`.

I agree. 
Actually, `DumpTensorShape::run` isn't problem for this.
In `dredd-rule-lib/rule-lib.sh`,
```
ACTUAL=`init_error_log ; \
          ${INSPECT_PROG_PATH} --tensor_shape ${COMPILED_FILE} | \
          awk -v tensor_name="$1" '{ if ($1 == tensor_name) print $2}'`
```

this line has problem.  To properly fix rule with blank, we have to add `for-loop` in `.sh` which seems too much for me.

Comment (icodo98) [reply res/TensorFlowLiteRecipes/Pad_002/test.rule:6]: How about `-1,5,7,2`?

**Comment by icodo98** [comment]
@nnfw-bot test nncc-debug nncc-release

File: res/TensorFlowLiteRecipes/Pad_002/test.recipe [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# padding with dynamic shape, others same as Pad_000
```
[the remark this answers is not in the document]
Comment (icodo98) [reply res/TensorFlowLiteRecipes/Pad_002/test.recipe:1]: > How about changing `Pad_002` directory to `Inf_Pad_00x` to separate it from other tests like `Net_Conv_Add_00x` ?

Nice idea!

I'll change it.

File: compiler/circle-inspect/src/Dump.cpp [hunk 3]
```diff
@@ -240,3 +240,38 @@ void DumpConstants::run(std::ostream &os, const circle::Model *model, const std:
...
+  {
+    reader.select_subgraph(g);
+    auto tensors = reader.tensors();
+
+    for (uint32_t i = 0; i < tensors->size(); ++i)
+    {
+      const auto tensor = tensors->Get(i);
+      auto shape = tensor->shape_signature() ? tensor->shape_signature() : tensor->shape();
+      os << reader.tensor_name(tensor) << " ";
+      for (int8_t i = 0; i < shape->size(); i++)
```
Comment (shs-park) [inline compiler/circle-inspect/src/Dump.cpp:264]: It's not a must, but most machines do calculations in 32-bit units, so there's probably no need to reduce it to 8-bit in `for loop`.

```suggestion
      for (uint32_t i = 0; i < shape->size(); i++)
```

File: res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# To check if dynamic dimension properly infered
+
+RULE    "VERIFY_FILE_FORMAT"      $(verify_file_format) '=' 1
+
+RULE    "PAD_EXIST"               $(op_count PAD) '=' 1
+RULE    "PAD_SHAPE"               $(tensor_shape ofm) '=' 1,-1,7,2
```
Comment (shs-park) [inline res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule:6]: The 'shape' expression `1,-1,7,2` is fine now, but it would be better to express it in a more formalized form like `[1, -1, 7, 2]`.
It would be better if it could handle spaces between commas as well.

```suggestion
RULE    "PAD_SHAPE"               $(tensor_shape ofm) '=' [1, -1, 7, 2]
```

**Review by shs-park — COMMENTED** [review]
It seems to work well!
I left a few minor comments.

File: res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# To check if dynamic dimension properly infered
```
Comment (shs-park) [inline res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule:1]: typo
=)

```suggestion
# To check if dynamic dimension properly inferred
```

File: compiler/circle2circle-dredd-recipe-test/test.lst [hunk 6]
```diff
@@ -96,6 +96,7 @@ Add(PadV2_001 PASS substitute_padv2_to_pad)
 Add(Softmax_001 PASS decompose_softmax)
 Add(Softmax_002 PASS decompose_softmax)
 Add(StridedSlice_003 PASS substitute_strided_slice_to_reshape)
+Add(Inf_Pad_000 PASS)
```
Comment (shs-park) [inline compiler/circle2circle-dredd-recipe-test/test.lst:99]: Please sort in alphabetical order

File: res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule (hunk 4 above)
Comment (icodo98) [reply res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule:6]: from #https://github.com/Samsung/ONE/issues/13998#issuecomment-2370601917

> FYI) RULE "PAD_SHAPE" $(tensor_shape ofm) '=' [1, -1, 7, 2] does not work with the current code. Currently, the rule only works if the arguments have exactly 4 elements, so there cannot be any spaces in between, such as [1, -1, 7, 2]. Therefore, at the moment, we are checking shapes in the form of 1,-1,7,2.
> 
> * Current code
>   https://github.com/Samsung/ONE/blob/646c7f16511f9baa8384f719e65cba768d163f7b/compiler/dredd-rule-lib/rule-lib.sh#L85-L92

File: res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule (hunk 5 above)
Comment (icodo98) [reply res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule:1]: S

> typo =)

Sorry, I'll fix it.

File: compiler/circle2circle-dredd-recipe-test/test.lst (hunk 6 above)
Comment (icodo98) [reply compiler/circle2circle-dredd-recipe-test/test.lst:99]: > Please sort in alphabetical order

Instead of sorting, what about separate `Inf_` tests? 

```suggestion
# Shape Inference test
Add(Inf_Pad_000 PASS)
```

File: compiler/circle-inspect/src/Dump.cpp (hunk 3 above)
Comment (icodo98) [reply compiler/circle-inspect/src/Dump.cpp:264]: > It's not a must, but most machines do calculations in 32-bit units, so there's probably no need to reduce it to 8-bit in `for loop`.

I'll update it.

File: res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule (hunk 4 above)
Comment (shs-park) [reply res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule:6]: Oh, thanks for the description.
How about using a quoted string, i.e., `"[1, -1, 7, 2]"`?
If it's difficult, the current implementation is not bad either =)

File: compiler/circle2circle-dredd-recipe-test/test.lst (hunk 6 above)
Comment (shs-park) [reply compiler/circle2circle-dredd-recipe-test/test.lst:99]: > , what about separate `Inf_` tests?

## 👍

File: res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# To check if dynamic dimension properly inferred
+
+RULE    "VERIFY_FILE_FORMAT"      $(verify_file_format) '=' 1
+
+RULE    "PAD_EXIST"               $(op_count PAD) '=' 1
```
Comment (kyeong8139) [inline res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule:5]: IMHO, `PAD_EXIST` is not related to SHAPE_INFERENCE.
Why is this necessary?

Comment (icodo98) [reply res/TensorFlowLiteRecipes/Inf_Pad_000/test.rule:5]: > Why is this necessary?

Your right.
